core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp

```diff
@@ -364,13 +364,19 @@ class TeamPolicyInternal<Kokkos::Experimental::SYCL, Properties...>
   template <class FunctorType>
   int internal_team_size_recommended_for(const FunctorType& f) const {
     // FIXME_SYCL improve
-    return internal_team_size_max_for(f);
+    const int max_team_size_half = internal_team_size_max_for(f) / 2;
+    int power_of_two             = 1;
+    while (power_of_two <= max_team_size_half) power_of_two <<= 1;
+    return power_of_two;
   }
 
   template <class FunctorType>
   int internal_team_size_recommended_reduce(const FunctorType& f) const {
     // FIXME_SYCL improve
-    return internal_team_size_max_reduce(f);
+    const int max_team_size_half = internal_team_size_max_reduce(f) / 2;
+    int power_of_two             = 1;
```

Comment:
There is this trick which makes the power of 2 calculation O(1) and no jumps:  [https://graphics.stanford.edu/~seander/bithacks.html#RoundUpPowerOf2](https://graphics.stanford.edu/~seander/bithacks.html#RoundUpPowerOf2)

Reply:
I'm using `Kokkos::log2` instead (which uses intrinsics where available).